### PR TITLE
Add DeepSeek-V3 NVFP4 training support

### PR DESCRIPTION
**Pull request by denys-fridman**
Squash of the dsv3_nvfp4 branch by Selvaraj Anandaraj (with contributions by Xiaowei Ren), rebased from its 26.04-alpha.rc2 base onto main. Commits already upstreamed to main (te_quant module-init context, name-plumbing None guards) were dropped during the rebase.

Contents:

- 1F1B EP-overlap schedule: replace the fp8-only inner context with get_low_precision_context(), which selects get_fp8_context() or get_fp4_context() per layer (fp4 skipped for MTP layers until MTP gained FP4 support below) so NVFP4 runs can capture the fine-grained overlap schedule under quantized contexts.
- MTP: build MultiTokenPredictionLayer under the FP4 init context and use FP4 in the 1F1B schedule for MTP layers.
- MLA: delayed wgrad compute support when the MLA down-projection GEMM fusion is enabled.
- Distributed optimizer checkpoint compatibility for fused MLA: when reloading main params from a state dict, match runtime parameter names against canonical checkpoint prefixes declared in sharded_state_dict_keys_map (e.g. fused-LN MLA stores LN scale/bias on linear_qkv_down_proj at runtime while checkpoints canonicalize them to input_layernorm.*).
- Selective per-module precision: thread module instance names from GPTModel into TransformerBlock ("decoder", "mtp") and down to layers ("decoder.layers.N") so TE quantization configs can target individual modules by name.

- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/transformer/multi_token_prediction.py
Comment (zhongbozhu) [inline megatron/core/transformer/multi_token_prediction.py]: Here^

File: megatron/core/transformer/transformer_block.py [hunk 1]
```diff
@@ -368,6 +370,11 @@ def build_layer(layer_spec, layer_number):
                     layer_number=layer_number,
                     pg_collection=self.pg_collection,
                     vp_stage=self.vp_stage,
+                    name=(
```
Comment (zhongbozhu) [inline megatron/core/transformer/transformer_block.py:373]: layer_number is 1-based (i + 1) for some reason (long time ago), while ModuleList paths are 0-based. The first MTP layer is registered as mtp.layers.0 but receives mtp.layers.1, causing per-module quantization recipes to miss or target the wrong layer.

File: megatron/core/optimizer/distrib_optimizer.py [hunk 2]
```diff
@@ -3069,13 +3069,50 @@ def _build_model_param_to_state_dict_param_map(self, state_dict):
             self._normalize_state_dict_for_grouped_params(state_dict_list[chunk_idx], model_chunk)
             self._synthesize_state_dict_params_for_model(state_dict_list[chunk_idx], model_chunk)
             names_in_state_dict = set(state_dict_list[chunk_idx].keys())
+
+            # Some layer specs declare a sharded_state_dict_keys_map that rewrites runtime
```
Comment (zhongbozhu) [inline megatron/core/optimizer/distrib_optimizer.py:3073]: why we didn't have this issue when doing mxfp8 DSv3? this is only needed for `--mla-down-proj-fusion` is that right?

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

File: megatron/core/optimizer/distrib_optimizer.py [hunk 3]
```diff
@@ -3069,13 +3069,50 @@ def _build_model_param_to_state_dict_param_map(self, state_dict):
...
+            # name against either its raw form or its canonical alternative.
+            canonical_rewrites = []
+            for layer_path, layer_module in model_chunk.named_modules():
+                cfg = getattr(layer_module, 'submodules_config', None)
+                if cfg is None:
+                    continue
+                key_map = getattr(cfg, 'sharded_state_dict_keys_map', None)
+                if not key_map:
+                    continue
+                layer_prefix = f"{layer_path}." if layer_path else ""
```
Comment (zhongbozhu) [inline megatron/core/optimizer/distrib_optimizer.py:3089]: model_chunk is the DDP wrapper here, so named_modules() returns paths such as module.decoder.layers.0. However, the parameter loop below strips the leading module. from name. This makes name.startswith(layer_prefix) false, so the sharded_state_dict_keys_map rewrite is never applied. Fused-LN MLA parameters will then fail to match canonical input_layernorm.* checkpoint keys when --load-main-params-from-ckpt is used. Please normalize layer_path with the same module.-stripping logic before constructing layer_prefix, and add a DDP-wrapped fused-MLA regression test.

File: megatron/core/optimizer/distrib_optimizer.py [hunk 4]
```diff
@@ -3069,13 +3069,50 @@ def _build_model_param_to_state_dict_param_map(self, state_dict):
...
+                        if sub.startswith(old_prefix):
+                            candidates.append(
+                                f"{layer_prefix}{new_prefix}{sub[len(old_prefix) :]}"
+                            )
+                            break
+                return candidates
+
             for name, model_param in model_chunk.named_parameters():
                 while name.startswith("module."):
                     name = name[len("module.") :]
```
Comment (zhongbozhu) [inline megatron/core/optimizer/distrib_optimizer.py:3108]: Here^, the "module" prefix get removed.
```
# before 
layer_prefix = "module.decoder.layers.0."  
# after 
name = "decoder.layers.0...."              
```

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

File: megatron/core/models/common/model_chunk_schedule_plan.py [hunk 5]
```diff
@@ -219,18 +220,22 @@ def set_fsdp_reshard_hooks(self, post_forward_hook, post_backward_hook):
...
-            get_fp8_context(self.layer.config, self.layer.layer_number - 1)
-            if use_inner_fp8_context
-            else nullcontext()
-        )
+        use_inner_fp4_context = self.layer.config.fp4
+
+        if use_inner_fp8_context:
+            low_precision_context = get_fp8_context(self.layer.config, self.layer.layer_number - 1)
+        elif use_inner_fp4_context:
+            low_precision_context = get_fp4_context(self.layer.config, self.layer.layer_number - 1)
```
Comment (zhongbozhu) [inline megatron/core/models/common/model_chunk_schedule_plan.py]: first_last_layers_bf16 applies to decoder layers, but this path passes an MTP depth as a decoder index. Normal/init paths currently quantize MTP, while overlap may leave it in BF16. Please choose one MTP precision policy and apply it consistently across initialization, normal forward, and overlap; do not derive it from MTP.layer_number.

File: megatron/core/models/common/model_chunk_schedule_plan.py [hunk 6]
```diff
@@ -219,18 +220,22 @@ def set_fsdp_reshard_hooks(self, post_forward_hook, post_backward_hook):
         # After the last forward op, release forward-pass params.
         last_fwd_node.set_post_forward_hook(lambda: post_forward_hook(hook_module))
 
-    def get_fp8_context(self):
+    def get_low_precision_context(self):
```
Comment (zhongbozhu) [inline megatron/core/models/common/model_chunk_schedule_plan.py:224]: suggested fix, essentially making sure first last bf16 is NOT affecting MTP but keeps regulating decoders. MTP layers do not inherit the layer number of decoder layers
```
# MTP layer numbers are depth-local, so exclude them from the decoder-only
# first/last BF16 policy and keep MTP quantized.
layer_no = -1 if self.is_mtp else self.layer.layer_number - 1

if use_inner_fp8_context:
    return get_fp8_context(self.layer.config, layer_no)
if use_inner_fp4_context:
    return get_fp4_context(self.layer.config, layer_no)
return nullcontext()
```

File: megatron/core/models/common/model_chunk_schedule_plan.py (hunk 5 above)
Comment (zhongbozhu) [reply megatron/core/models/common/model_chunk_schedule_plan.py]: You can see that the `multi_token_prediction.py` without 1f1b overlap has the correct code:

```
       if self.config.fp8:
            quantization_context = get_fp8_context(self.config)  # layer_no=-1
            transformer_layer_quantization_context = get_fp8_context(self.config)
        elif self.config.fp4:
            quantization_context = get_fp4_context(self.config)  # layer_no=-1
            transformer_layer_quantization_context = get_fp4_context(self.config)
```

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

File: megatron/core/transformer/transformer_block.py (hunk 1 above)
Comment (denys-fridman) [reply megatron/core/transformer/transformer_block.py:373]: fixed

File: megatron/core/optimizer/distrib_optimizer.py (hunk 2 above)
Comment (denys-fridman) [reply megatron/core/optimizer/distrib_optimizer.py:3075]: Yep, this is only needed with `--mla-down-proj-fusion`

File: megatron/core/optimizer/distrib_optimizer.py (hunk 3 above)
Comment (denys-fridman) [reply megatron/core/optimizer/distrib_optimizer.py:3091]: Fixed in e64a86a41: the module path is now stripped with the same `while name.startswith("module.")` loop before building layer_prefix. Also added the test

File: megatron/core/optimizer/distrib_optimizer.py (hunk 4 above)
Comment (denys-fridman) [reply megatron/core/optimizer/distrib_optimizer.py:3110]: done

File: megatron/core/models/common/model_chunk_schedule_plan.py (hunk 5 above)
Comment (denys-fridman) [reply megatron/core/models/common/model_chunk_schedule_plan.py]: done

File: megatron/core/models/common/model_chunk_schedule_plan.py (hunk 6 above)
Comment (denys-fridman) [reply megatron/core/models/common/model_chunk_schedule_plan.py:219]: done

**Review by zhongbozhu — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by guihong-nv — APPROVED** [review]
